test/e2e/framework/nodes.go

```diff
@@ -59,6 +59,32 @@ func findNodesByGatewayLabel(cluster int, isGateway bool) []*v1.Node {
 	return retNodes
 }
 
+func (f *Framework) FindMultipleGatewaysCluster(clusters []string) ([]int, bool) {
```

Comment:
So this function finds one cluster with multiple GWs and returns it in an array as the first element. I assume this is so the caller can easily identify some other cluster if needed. However this wasn't intuitive to me at first. Perhaps we can simplify it to find and return just the cluster index with multiple GWs (or -1 if not found). If a caller needs to identify some other cluster then we can provide a helper to do so.

```suggestion
func (f *Framework) FindClusterWithMultipleGateways(clusters []string) int {
```

Reply:
We really should pass the cluster indices to search otherwise the indices of the local `clusters` array may not align with the global indices `findNodesByGatewayLabel` expects.

That said, why not search all the globally defined clusters or will there be a use case where the caller will want to search a subset?

Reply:
@tpantelis 

Ok, I'm trying to follow.
You're saying that it's best to return an index of the cluster or if not clusters found, return "-1".

This is how I'm calling the function:
`clustersIndex, gatewaysExist := f.FindMultipleGatewaysCluster(framework.TestContext.ClusterIDs)`

Then I'm taking the first cluster from the list - `clusterIndex[0]`.
And since I'm in any case need only the first cluster, I can return only the first found cluster index.

Am I got it right?

Reply:
Yes - returning an array with the remaining clusters as well is extraneous.

Reply:
Great, will make the changes.
Thanks!

Reply:
Hi @tpantelis 

Will appreciate another advice, please.
We agreed that the `FindMultipleGatewaysCluster` function will return an index of the single cluster with multiple GW nodes.

But I need identify the second cluster as well.
Meaning, if the cluster with 2 GW nodes has an index 1, I need to define the second cluster with index 0.
For the second cluster, it doesn't matter for me, what the amount of the GW nodes it has.

So, I thought about two possible solutions.
1) Create a function `findGatewayFailoverClusters`. That function will call for the `FindMultipleGatewaysCluster` in turn and when get the index of the cluster with multiple GW nodes, will loop through the "framework.TestContext.ClusterIDs" to find a second cluster which does not match the returned cluster index. Then it will return two variables. One for primaryCluster and one secondaryCluster.
2) Perform the loop within the body of the test to define the second cluster.

What do you think is preferred?
Or maybe any other way?

Thank you.

Reply:
> Will appreciate another advice, please. We agreed that the `FindMultipleGatewaysCluster` function will return an index of the single cluster with multiple GW nodes.

Actually I notice in https://github.com/submariner-io/submariner/pull/2014 that you also need the GW node instances so rather than querying for those again we might as well return them here as well.

> 
> But I need identify the second cluster as well. Meaning, if the cluster with 2 GW nodes has an index 1, I need to define the second cluster with index 0. For the second cluster, it doesn't matter for me, what the amount of the GW nodes it has.
> 
> So, I thought about two possible solutions.
> 
> 1. Create a function `findGatewayFailoverClusters`. That function will call for the `FindMultipleGatewaysCluster` in turn and when get the index of the cluster with multiple GW nodes, will loop through the "framework.TestContext.ClusterIDs" to find a second cluster which does not match the returned cluster index. Then it will return two variables. One for primaryCluster and one secondaryCluster.
> 2. Perform the loop within the body of the test to define the second cluster.
> 
> What do you think is preferred? Or maybe any other way?
> 

I suggest #2. You could add a helper function called FindOtherClusterIndex (put it in _util.go_).

Reply:
Cool, thanks!

Reply:
@tpantelis 

A question regarding the helper function, please.

If I'm adding a function to util.go within shipyard repo, it does not available from the GatewayFailover test in submariner repo.
The framework module already imported.
What am I missing?

Thank you.

Reply:
Is the new function exported from the package, ie capitalized?

Reply:
Yes.
It looks like the following:

`
func FindOtherClusterIndex(initialCluster int) {
	...func body...
}
`

Reply:
If you want to use a new function provided by Shipyard in Submariner, you need to update the dependency on Shipyard (`go get github.com/submariner-io/shipyard@devel` once the PR is merged). To test locally, you need to add a local replace: `replace github.com/submariner-io/shipyard => ../shipyard` in Submariner’s `go.mod` (but make sure you don’t commit that).

Reply:
@skitt 

Yes, I know.
I'm already using that way for local testing.
The strange thing is that any new function within the `framework.go`, `gateways.go` or `nodes.go` under shipyard, gets available in the submariner repo (with the "replace").
But the function under `util.go` is not recognized.

Of course, I;m doing a `go mod vendor`after any change in shipyard repo.

Reply:
After your `go mod vendor`, is `vendor/github.com/submariner-io/shipyard/test/e2e/framework/util.go` in your Submariner repo updated to include your changes in Shipyard?

Reply:
Yes, I can see it.

But I still go the following error.

```
# github.com/submariner-io/submariner/test/e2e/redundancy
test/e2e/redundancy/gateway_failover.go:148:23: f.FindOtherClusterIndex undefined (type *"github.com/submariner-io/submariner/test/e2e/framework".Framework has no field or method FindOtherClusterIndex)
note: module requires Go 1.18
FAIL	github.com/submariner-io/submariner/test/e2e [build failed]
FAIL
```

Reply:
I also thought that the problem is that I used go 1.17.
I upgraded to go 1.18.6, and the error still the same.

Reply:
Ah, you’re calling the function on a framework, but it’s not declared as a `Framework` function. If `FindOtherClusterIndex` needs information from `Framework`, you need to declare it as

    func (f *Framework) FindOtherClusterIndex(…) …

If it doesn’t, you need to call it as `framework.FindOtherClusterIndex(…)`, not `f.FindOtherClusterIndex(…)`.

Reply:
Right!
I forgot that I already seen similar issue.

@skitt Thanks a lot for showing me the mistake.